### PR TITLE
Fix issue 1474 (to check invalid sheet name)

### PR DESCRIPTION
**Pull request by skypesky**
## Summary

Fix issue #1474 .
An exception is thrown when the name is illegal.

## Test plan

For this I have added unit tests.

### CODE REVIEW

**Comment by papb** [comment]
Hello! I suggest you take a look [at this](https://github.com/papb/json-excel/blob/master/source/assert-valid-sheet-names.ts), do you think it would be good to also add these other checks?

**Comment by skypesky** [comment]
hi,papb,Your method is more concise, I think it's very good.
I didn't write according to the assertvalidsheetname method before, because the validation of name in the original code did not always throw an exception. In this way, how to organize it becomes very strange.

About double quotation marks
---
It seems that the naming rules of sheet have something to do with the version of Excel

As for single quotation marks
---
As far as I know, single quotation marks can be placed in the middle, but not the head and tail. Can you test the single quotation marks for me?

File: spec/integration/worksheet.spec.js [hunk 1]
```diff
@@ -659,6 +659,22 @@ describe('Worksheet', () => {
       });
     });
 
+    context('when the worksheet name contains illegal characters', () => {
+      it('when the worksheet name contain < > * ? " | : / \\ [ ]', () => {
```
Comment (alubbe) [inline spec/integration/worksheet.spec.js:663]: this should be `it('throws an error'...)`

File: spec/integration/worksheet.spec.js [hunk 2]
```diff
@@ -659,6 +659,22 @@ describe('Worksheet', () => {
...
+        const workbook = new ExcelJS.Workbook();
+
+        const invalideNames = ['*', '?', ':', '/', '\\', '[', ']'];
```
Comment (alubbe) [inline spec/integration/worksheet.spec.js:666]: please call this `invalidCharacters`

File: lib/doc/workbook.js [hunk 3]
```diff
@@ -97,6 +102,12 @@ class Workbook {
     return worksheet;
   }
 
+  // Illegal character in worksheet name: asterisk (*), question mark (?),
+  // colon (:), forward slash (/ \), or bracket ([])
+  containIllegalSheetName(sheetName) {
+    return /(\*)|(\?)|(:)|(\/)|(\\)|(\[)|(\])/.test(sheetName);
```
Comment (alubbe) [inline lib/doc/workbook.js:108]: please use the simpler and faster regex `/[*?:\/\\\[\]]/` instead

**Review by alubbe — CHANGES_REQUESTED** [review]

File: spec/integration/worksheet.spec.js [hunk 4]
```diff
@@ -659,6 +659,48 @@ describe('Worksheet', () => {
...
+            workbook.addWorksheet(invalidCharacters[lastIndex])
+          ).to.throw(expectedErrors[lastIndex]);
+        }
+      });
+
+      it('throws an error', () => {
+        const workbook = new ExcelJS.Workbook();
+
+        const invalidNames = ['\'sheetName', 'sheetName\''];
+        const expectedErrors = invalidNames.map(
```
Comment (alubbe) [inline spec/integration/worksheet.spec.js:687]: I think the array creation and the for-index-loop distract a little bit from what is being tested, I think we should simplify to this:
```js
for (const invalidName of invalidNames) {
  expect(() => workbook.addWorksheet(invalidName)).to.throw(
    `The first or last character of worksheet name cannot be a single quotation mark: ${invalidName}`
  );
}
```

File: spec/integration/worksheet.spec.js [hunk 5]
```diff
@@ -659,6 +659,48 @@ describe('Worksheet', () => {
       });
     });
 
+    context('when the worksheet name contains illegal characters', () => {
+      it('throws an error', () => {
+        const workbook = new ExcelJS.Workbook();
+
+        const invalidCharacters = ['*', '?', ':', '/', '\\', '[', ']'];
+        const expectedErrors = invalidCharacters.map(
```
Comment (alubbe) [inline spec/integration/worksheet.spec.js:667]: I think the array creation and the for-index-loop distract a little bit from what is being tested, I think we should simplify to this:
```js
for (const invalidCharacter of invalidCharacters) {
  expect(() => workbook.addWorksheet(invalidCharacter)).to.throw(
    `Worksheet name ${invalidCharacter} cannot include any of the following characters: * ? : \\ / [ ]`
  );
}
```

**Review by alubbe — CHANGES_REQUESTED** [review]
Just some minor feedback around the readability/maintainability of the test case, then we're good to go!

**Review by alubbe — APPROVED** [review]
thank you!

**Comment by Sczlog** [comment]
Although here has been a long long times since this change applied.
But I suppose if replace invalid character with `_` should be a better choice?
Current solution is kinda a breaking change, code running successful (but not correct?) before will break after update to later version.

**Pull request merged**